google/cloud/internal/credentials.h

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_CREDENTIALS_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_CREDENTIALS_H
+
+#include "google/cloud/version.h"
+#include "absl/types/optional.h"
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+class CredentialsVisitor;
+}  // namespace internal
+
+// TODO(#6293) - move unified credentials out of internal namespace
+namespace internal {
+class Credentials {
+ public:
+  virtual ~Credentials() = 0;
+
+ private:
+  friend class CredentialsVisitor;
+  virtual void dispatch(CredentialsVisitor& visitor) = 0;
+};
+
+std::shared_ptr<Credentials> GoogleDefaultCredentials();
+std::shared_ptr<Credentials> AccessTokenCredentials(
+    std::string const& access_token,
+    std::chrono::system_clock::time_point expiration);
+
+/// A wrapper to store credentials into an options
+struct UnifiedCredentialsOption {
+  using Type = std::shared_ptr<Credentials>;
+};
+
+}  // namespace internal
+
+namespace internal {
+
+/// Represents an access token with a known expiration time.
+struct AccessToken {
+  std::string token;
+  std::chrono::system_clock::time_point expiration;
+};
+
+/// A generator for access tokens.
+using AccessTokenSource = std::function<AccessToken()>;
+
+/// A generic generator of access tokens.
+std::shared_ptr<Credentials> DynamicAccessTokenCredentials(
```

Comment:
Do you think these factories, all of which return a `Credentials` (shared_ptr), would make sense all named `MakeCredentials()` and overloaded based on the argument. Like WDYT about

```cc
std::shared_ptr<Credentials> MakeCredentials();  // Retruns Google Default Credentials
std::shared_ptr<Credentials> MakeCredentials(AccessToken const& tok);
std::shared_ptr<Credentials> MakeCredentails(AccessTokenSource source);
...
```

Reply:
With respect to the `Make*` vs. not, I have very little opinion, I think `Make*` may avoid ambiguity because too many functions are called `GoogleDefaultCredentials()` already (in different namespaces). I will make that change.

There will be at least a few that take no arguments, e.g., `MakeComputeEngineCredentials()` and `MakeGoogleDefaultCredentials()`.  Some will take only a string, which of course we could wrap in a struct to make the overload works.  But that seems like too much wrapping.

Reply:
A leading `Make` SGTM. Avoiding overloading for the reasons you mentioned SGTM.